internal/pally/registry.go

```diff
@@ -0,0 +1,250 @@
+package pally
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"github.com/prometheus/common/model"
+	"github.com/uber-go/tally"
+	"go.uber.org/atomic"
+)
+
+// A Registry is a collection of metrics, usually scoped to a single package or
+// object. Each Registry is also its own http.Handler, and can serve
+// Prometheus-flavored text and protocol buffer pages for metrics introspection
+// or scraping.
+type Registry struct {
+	http.Handler
+
+	metricsMu   sync.RWMutex
+	metrics     []metric
+	constLabels Labels
+
+	// TODO: To serve our own Prometheus endpoints, we only need to implement
+	// prometheus.Gatherer. The vanilla Prometheus registry is a convenient
+	// implementation, but it enforces metric uniqueness via FNV64a; if we run
+	// into hash collisions in our more-demanding applications, we should use a
+	// simple map[string]struct{} instead.
+	prom      *prometheus.Registry
+	federated []prometheus.Registerer
+
+	pushing atomic.Bool
+	scope   tally.Scope
+	ticker  *time.Ticker
+	stop    chan struct{}
+	stopped chan struct{}
+}
+
+// A RegistryOption configures a Registry.
+type RegistryOption func(*Registry)
```

Comment:
optional: it's useful to wrap option types in a struct so that users can't implement their own

```
type RegistryOption struct{ apply func(*Registry) }
```

Otherwise, since Registry is public and it publicly embeds `http.Handler`, the following is a valid option:

```
func nilHandler(r *Registry) { r.Handler = nil }
```